core/src/main/java/org/elasticsearch/index/cache/bitset/BitsetFilterCache.java

```diff
@@ -95,16 +100,17 @@ public BitsetFilterCache(Index index, @IndexSettings Settings indexSettings) {
     @Inject(optional = true)
     public void setIndicesWarmer(IndicesWarmer indicesWarmer) {
         this.indicesWarmer = indicesWarmer;
+        indicesWarmer.addListener(warmer);
     }
 
-    public void setIndexService(IndexService indexService) {
-        this.indexService = indexService;
-        // First the indicesWarmer is set and then the indexService is set, because of this there is a small window of
-        // time where indexService is null. This is why the warmer should only registered after indexService has been set.
-        // Otherwise there is a small chance of the warmer running into a NPE, since it uses the indexService
-        indicesWarmer.addListener(warmer);
+    public void setListener(Listener listener) {
+        if (listener == null) {
+            throw new IllegalArgumentException("listener must not be null");
+        }
+        this.listener = listener;
```

Comment:
I think it would help catch mistakes to make sure that we are not overriding a listener that was already set. Maybe we could make the default listener a constant and throw an exception here if the current listener is not the default one, meaning that two different components tried to set a listener? Otherwise only the last setListener call would win, which could create bugs that are hard to debug?